File: fast_sync_dart_client/fast_sync_client/lib/src/implementation/service/sync_manager.dart
import 'package:fast_sync_client/fast_sync_client.dart';

class SyncManager implements ISyncManager {
  @override
  Future<SyncPayload> push() async {
    IhttpManager httpManager = FastSync.getHttpManager();

    SyncPayload payload = await _buildPayload();
    bool isSucced = false;

    if (payload.hasData) {
      isSucced = await httpManager.push(payload);
    }
    if (isSucced) {
      await _undirtyList(payload);
    }
    return payload;
  }

  @override
  Future<SyncPayload> pull() async {
    IhttpManager httpManager = FastSync.getHttpManager();
    SyncOperationMetadata operationMetadata = SyncOperationMetadata();
    List<String> syncableTypes = FastSync.getSyncableTypes();
    for (String type in syncableTypes) {
      SyncVersionManager syncVersionManager = FastSync.getSyncVersionManager();
      ISyncMetadata typeLocalMetadata =
          await syncVersionManager.getTypeSyncMetadata(type);
      operationMetadata.setMetadata(type, typeLocalMetadata);
    }

    SyncPayload payload = await httpManager.pull(operationMetadata);
    await _undirtyList(payload);
    await _processPayloadMetadata(payload);
    return payload;
  }

  @override
  Future<SyncPayload<ISyncableObject>> hardReset({List<Type>? types}) async {
    List<String> syncableTypes = [];
    if (types != null && types.isNotEmpty) {
      for (Type type in types) {
        syncableTypes.add(type.toString());
      }
    } else {
      syncableTypes = FastSync.getSyncableTypes();
    }
    for (String type in syncableTypes) {
      ISyncableRepository<ISyncableObject> repository =
          FastSync.getObjectRepository(type: type);
      await repository.hardDelete();
    }
    return await pull();
  }

  bool _filterDirtyObjects(ISyncableObject object) {
    return object.dirty;
  }

  Future<SyncPayload> _buildPayload() async {
    List<String> syncableTypes = FastSync.getSyncableTypes();
    SyncPayload payload = SyncPayload();
    for (String type in syncableTypes) {
      ISyncableRepository<ISyncableObject> repository =
          FastSync.getObjectRepository(type: type);
      List<ISyncableObject> dirtyObjects =
          await repository.query(_filterDirtyObjects);
      payload.pushObjects(type, dirtyObjects);
    }
    return payload;
  }

  Future<void> _undirtyList(SyncPayload payload) async {
    List<String> syncableTypes = payload.getSyncedTypes();
    for (String type in syncableTypes) {
      ISyncableRepository<ISyncableObject> repository =
          FastSync.getObjectRepository(type: type);
      List<ISyncableObject> pushedItems = payload.getObjectsForType(type);
      await repository.undirtyList(pushedItems);
    }
  }

  Future<void> _processPayloadMetadata(SyncPayload payload) async {
    List<String> syncableTypes = payload.getSyncedTypes();
    SyncVersionManager syncVersionManager = FastSync.getSyncVersionManager();
    for (String type in syncableTypes) {
      ISyncMetadata metadata = payload.getTypeMetadata(type);
      await syncVersionManager.updateTypeSyncVersion(metadata);
    }
  }
}

File: fast-sync-typescript-client-sdk/packages/fast-sync-client/src/implementation/service/sync_manager.ts
import { ISyncableRepository } from "../../abstraction/data/isyncable_repository";
import { ISyncMetadata } from "../../abstraction/metadata/isync_metadata";
import { ISyncableObject } from "../../abstraction/metadata/isyncable_object";
import { SyncOperationMetadata } from "../../abstraction/models/sync_operation_metadata";
import { SyncPayload } from "../../abstraction/models/sync_payload";
import { IHttpManager } from "../../abstraction/service/ihttp_manager";
import { ISyncManager } from "../../abstraction/service/isync_manager";
import { FastSync } from "../fast_sync";
import { SyncVersionManager } from "./sync_version_manager";

export class SyncManager implements ISyncManager {
  async push(): Promise<SyncPayload> {
    const httpManager: IHttpManager = FastSync.getHttpManager();
    const payload: SyncPayload = await this._buildPayload();
    let isSucced = false;

    if (payload.hasData) {
      isSucced = await httpManager.push(payload);
    }
    if (isSucced) {
      await this._undirtyList(payload);
    }
    return payload;
  }

  async pull(): Promise<SyncPayload> {
    const httpManager: IHttpManager = FastSync.getHttpManager();
    const operationMetadata = new SyncOperationMetadata();
    const syncableTypes: string[] = FastSync.getSyncableTypes();

    for (const type of syncableTypes) {
      const syncVersionManager: SyncVersionManager =
        FastSync.getSyncVersionManager();
      const typeLocalMetadata: ISyncMetadata =
        await syncVersionManager.getTypeSyncMetadata(type);
      operationMetadata.setMetadata(type, typeLocalMetadata);
    }

    const payload: SyncPayload = await httpManager.pull(operationMetadata);
    await this._undirtyList(payload);
    await this._processPayloadMetadata(payload);
    return payload;
  }

  async sync(): Promise<SyncPayload> {
    await this.push();
    return await this.pull();
  }

  async hardReset(
    types?: Array<{ new (): ISyncableObject }>
  ): Promise<SyncPayload> {
    let syncableTypes: string[] = [];

    if (types && types.length > 0) {
      syncableTypes = types.map((type) => type.name);
    } else {
      syncableTypes = FastSync.getSyncableTypes();
    }

    for (const type of syncableTypes) {
      const repository: ISyncableRepository<ISyncableObject> =
        FastSync.getObjectRepository(type);
      await repository.hardDelete();
    }

    return await this.pull();
  }

  private _filterDirtyObjects(object: ISyncableObject): boolean {
    return object.dirty;
  }

  private async _buildPayload(): Promise<SyncPayload> {
    const syncableTypes: string[] = FastSync.getSyncableTypes();
    const payload: SyncPayload = new SyncPayload();

    for (const type of syncableTypes) {
      const repository: ISyncableRepository<ISyncableObject> =
        FastSync.getObjectRepository(type);
      const dirtyObjects: ISyncableObject[] = await repository.query(
        this._filterDirtyObjects
      );
      payload.pushObjects(type, dirtyObjects);
    }

    return payload;
  }

  private async _undirtyList(payload: SyncPayload): Promise<void> {
    const syncableTypes: string[] = payload.getSyncedTypes();

    for (const type of syncableTypes) {
      const repository: ISyncableRepository<ISyncableObject> =
        FastSync.getObjectRepository(type);
      const pushedItems: ISyncableObject[] = payload.getObjectsForType(type);
      const createFunction = FastSync.getTypeCreateFunction(type);
      let list = pushedItems.map((item) => createFunction(item));
      // for (var i = 0; i < pushedItems.length; i++) {
      //   list.push(createFunction(pushedItems[i]));
      // }
      await repository.undirtyList(list);
    }
  }

  private async _processPayloadMetadata(payload: SyncPayload): Promise<void> {
    const syncableTypes: string[] = payload.getSyncedTypes();
    const syncVersionManager: SyncVersionManager =
      FastSync.getSyncVersionManager();

    for (const type of syncableTypes) {
      const metadata: ISyncMetadata = payload.getTypeMetadata(type);
      await syncVersionManager.updateTypeSyncVersion(metadata);
    }
  }
}
